components/datetime/src/format/datetime.rs

```diff
@@ -573,6 +573,112 @@ impl RequiredData {
     }
 }
 
+/// What data is required to format a given pattern.
+#[derive(Default)]
+pub(crate) struct RequiredDataNeo {
+    // C::YearSymbolsV1Marker is required.
+    pub year_symbols: Option<Field>,
+    // C::MonthSymbolsV1Marker is required.
+    pub month_symbols: Option<Field>,
+    // WeekdaySymbolsV1Marker is required.
+    pub weekday_symbols: Option<Field>,
+    // DayPeriodSymbolsV1Marker is required.
+    pub dayperiod_symbols: Option<Field>,
+    // FixedDecimalFormatter is required.
+    pub fixed_decimal_formatter: bool,
+    // WeekDataV1 is required.
+    pub week_data: bool,
+}
+
+impl RequiredDataNeo {
+    #[inline]
+    pub(crate) fn needs_everything(&self) -> bool {
+        self.year_symbols.is_some()
+            && self.month_symbols.is_some()
+            && self.weekday_symbols.is_some()
+            && self.dayperiod_symbols.is_some()
+            && self.fixed_decimal_formatter
+            && self.week_data
+    }
+
+    /// Checks if formatting `pattern` would require us to load data & if so adds
+    /// them to this struct.
+    /// Keep it in sync with the `write_field` use of symbols.
+    pub(crate) fn add_requirements_from_pattern(
+        &mut self,
+        pattern: &Pattern,
+        supports_time_zones: bool,
+    ) -> Result<bool, Field> {
+        let fields = pattern.items.iter().filter_map(|p| match p {
+            PatternItem::Field(field) => Some(field),
+            _ => None,
+        });
+
+        enum NeedsWhat<'a> {
+            Nothing,
+            Symbols(&'a mut Option<Field>),
+            FixedDecimalFormatter,
```

Comment:
praise: nice, clean, and extensible

Reply:
Thanks, but I inlined this function directly into DateTimePatternInterpolator in f53809a which is not quite as elegant but it has fewer invariants between modules which I think is better overall.